File: lib/near/assets/test-suite/handler.ts
import { connect, keyStores, KeyPair, providers, transactions, utils } from 'near-api-js';
import { CloudWatchClient, PutMetricDataCommand, StandardUnit } from '@aws-sdk/client-cloudwatch';

// Configuration from environment
const RPC_URL = process.env.RPC_URL!;
const NETWORK_ID = process.env.NETWORK_ID || 'localnet';
const INCLUDE_WRITE_TESTS = process.env.INCLUDE_WRITE_TESTS === 'true';
const TEST_DEPTH = process.env.TEST_DEPTH || 'basic';

// Test results tracking
let testsPassed = 0;
let testsFailed = 0;

// CloudWatch client
const cloudwatch = new CloudWatchClient({});

async function putMetric(metricName: string, value: number, unit: StandardUnit = StandardUnit.Count): Promise<void> {
    try {
        await cloudwatch.send(new PutMetricDataCommand({
            Namespace: 'NEAR/Test',
            MetricData: [{
                MetricName: metricName,
                Value: value,
                Unit: unit,
                Dimensions: [{
                    Name: 'FunctionName',
                    Value: process.env.AWS_LAMBDA_FUNCTION_NAME || 'near-localnet-test'
                }]
            }]
        }));
    } catch (error) {
        console.error(`Warning: Failed to put metric ${metricName}:`, error);
    }
}

function log(message: string): void {
    const timestamp = new Date().toISOString();
    console.log(`[${timestamp}] ${message}`);
}

async function runTest(name: string, testFn: () => Promise<void>): Promise<boolean> {
    const startTime = Date.now();
    try {
        await testFn();
        const duration = Date.now() - startTime;
        testsPassed++;
        log(`✅ PASS: ${name} (${duration}ms)`);
        return true;
    } catch (error) {
        const duration = Date.now() - startTime;
        testsFailed++;
        log(`❌ FAIL: ${name} (${duration}ms)`);
        log(`Error: ${error instanceof Error ? error.message : String(error)}`);
        return false;
    }
}

async function runReadTests(provider: any): Promise<void> {
    log('=== Running Read-Only Tests ===');
    
    // Test 1: RPC Status
    await runTest('RPC Status', async () => {
        const status = await provider.status();
        if (!status.chain_id) {
            throw new Error('No chain_id in status');
        }
        log(`Chain ID: ${status.chain_id}`);
    });
    
    // Test 2: Latest Block
    await runTest('Latest Block', async () => {
        const block = await provider.block({ finality: 'final' });
        if (!block.header) {
            throw new Error('No block header');
        }
        log(`Block height: ${block.header.height}`);
    });
    
    // Test 3: Network Info
    await runTest('Network Info', async () => {
        const networkInfo = await provider.sendJsonRpc('network_info', []);
        if (!networkInfo) {
            throw new Error('No network info');
        }
        log(`Network info retrieved: ${JSON.stringify(networkInfo).substring(0, 100)}...`);
    });
    
    // Test 4: Query Account (node0 validator)
    await runTest('View System Account', async () => {
        const account = await provider.query({
            request_type: 'view_account',
            account_id: 'node0',
            finality: 'final'
        }) as any;
        if (!account || !account.amount) {
            throw new Error('Account not found or invalid');
        }
        log(`Account balance: ${utils.format.formatNearAmount(account.amount)} NEAR`);
    });
}

async function runWriteTests(
    provider: any,
    masterKey: KeyPair,
    masterAccountId: string,
    testDepth: string = TEST_DEPTH
): Promise<void> {
    log('=== Running Write Tests ===');
    
    // Create account ID and key pair outside test functions so they can be reused
    // Format: <random-hex>.node0 (subaccount of node0 validator for localnet)
    const randomSuffix = Math.random().toString(36).substring(2, 10);
    const testAccountId = `${randomSuffix}.node0`;
    const testKeyPair = KeyPair.fromRandom('ed25519');
    
    // Track if account creation succeeded for subsequent tests
    let accountCreated = false;
    
    // Test 1: Create Account
    const createAccountSuccess = await runTest('Create Test Account', async () => {
        // Connect to master account using near-api-js Account API
        const keyStore = new keyStores.InMemoryKeyStore();
        await keyStore.setKey(NETWORK_ID, masterAccountId, masterKey);
        
        const config = {
            networkId: NETWORK_ID,
            keyStore,
            nodeUrl: RPC_URL,
        };
        
        const near = await connect(config);
        const masterAccount = await near.account(masterAccountId);
        
        // Create subaccount with initial balance
        // For localnet, we can create a subaccount of node0 (e.g., test-123.test.near)
        const amount = utils.format.parseNearAmount('10')!;
        
        try {
            // Use Account.createAccount for simpler API
            const result = await masterAccount.createAccount(
                testAccountId,
                utils.PublicKey.from(testKeyPair.getPublicKey()),
                BigInt(amount)
            );
            
            log(`Account creation transaction: ${result.transaction.hash}`);
            
            // Wait for transaction to complete
            await new Promise(resolve => setTimeout(resolve, 3000));
            
            // Verify account was created
            const account = await provider.query({
                request_type: 'view_account',
                account_id: testAccountId,
                finality: 'final'
            }) as any;
            
            if (!account || !account.amount) {
                throw new Error('Account was not created successfully');
            }
            
            log(`Test account created: ${testAccountId} with balance ${utils.format.formatNearAmount(account.amount)} NEAR`);
        } catch (error: any) {
            log(`createAccount failed: ${error.message}`);
            // For write tests, we'll skip if account creation fails
            // This is acceptable as the read tests already passed
            throw error;
        }
        
        accountCreated = true;
    });
    
    // Test 2: Send Transaction (if account created)
    if (accountCreated && testDepth === 'comprehensive') {
        await runTest('Send Transaction', async () => {
            // Connect to the test account using the key pair we created
            const keyStore = new keyStores.InMemoryKeyStore();
            await keyStore.setKey(NETWORK_ID, testAccountId, testKeyPair);
            
            const config = {
                networkId: NETWORK_ID,
                keyStore,
                nodeUrl: RPC_URL,
            };
            
            const near = await connect(config);
            const testAccount = await near.account(testAccountId);
            
            // Transfer 1 NEAR back to node0
            const transferAmount = BigInt(utils.format.parseNearAmount('1')!);
            const result = await testAccount.sendMoney(
                masterAccountId,
                transferAmount
            );
            
            log(`Transfer transaction hash: ${result.transaction.hash}`);
            
            // Wait for transaction to complete
            await new Promise(resolve => setTimeout(resolve, 2000));
            
            // Verify transfer by checking balances
            const receiverAccount = await provider.query({
                request_type: 'view_account',
                account_id: masterAccountId,
                finality: 'final'
            }) as any;
            
            log(`Receiver balance: ${utils.format.formatNearAmount(receiverAccount.amount)} NEAR`);
        });
    }
}

export async function handler(event: any): Promise<any> {
    testsPassed = 0;
    testsFailed = 0;
    const startTime = Date.now();
    
    // Override from event if provided
    const includeWriteTestsFromEvent = typeof event === 'object' && event !== null && 'includeWriteTests' in event
        ? event.includeWriteTests === true || event.includeWriteTests === 'true'
        : INCLUDE_WRITE_TESTS;
    const testDepthFromEvent = typeof event === 'object' && event !== null && 'testDepth' in event
        ? event.testDepth
        : TEST_DEPTH;
    
    log('=== NEAR Localnet Test Suite ===');
    log(`RPC URL: ${RPC_URL}`);
    log(`Network ID: ${NETWORK_ID}`);
    log(`Include Write Tests: ${includeWriteTestsFromEvent}`);
    log(`Test Depth: ${testDepthFromEvent}`);
    
    try {
        // Validate RPC URL
        if (!RPC_URL) {
            throw new Error('RPC_URL environment variable not set');
        }
        
        // Create JSON RPC provider
        const provider = new providers.JsonRpcProvider({ url: RPC_URL });
        
        // Measure RPC response time
        const rpcStart = Date.now();
        await provider.status();
        const rpcDuration = Date.now() - rpcStart;
        log(`RPC response time: ${rpcDuration}ms`);
        
        // Send CloudWatch metric
        await putMetric('RpcResponseTime', rpcDuration, StandardUnit.Milliseconds);
        
        // Run read tests (always)
        await runReadTests(provider);
        
        // Run write tests (conditional, requires validator key)
        if (includeWriteTestsFromEvent) {
            // Get validator key from environment
            const validatorKeyJson = process.env.VALIDATOR_KEY_JSON;
            if (validatorKeyJson) {
                try {
                    const keyData = JSON.parse(validatorKeyJson);
                    const secretKey = keyData.secret_key || keyData.private_key;
                    if (!secretKey) {
                        throw new Error('Invalid key format: missing secret_key or private_key');
                    }
                    
                    // Handle both "ed25519:..." format and raw base64
                    let masterKey: KeyPair;
                    if (secretKey.startsWith('ed25519:')) {
                        masterKey = KeyPair.fromString(secretKey);
                    } else {
                        // Assume it's base64 encoded
                        masterKey = KeyPair.fromString(`ed25519:${secretKey}`);
                    }
                    
                    const masterAccountId = keyData.account_id || 'node0';
                    
                    log(`Using validator key for account: ${masterAccountId}`);
                    await runWriteTests(provider, masterKey, masterAccountId, testDepthFromEvent);
                } catch (error) {
                    log(`WARNING: Failed to parse validator key: ${error instanceof Error ? error.message : String(error)}`);
                    log('Skipping write tests due to invalid validator key');
                }
            } else {
                log('WARNING: VALIDATOR_KEY_JSON not set, skipping write tests');
            }
        }
        
        // Send metrics
        const totalDuration = Date.now() - startTime;
        await putMetric('TestsPassed', testsPassed);
        await putMetric('TestsFailed', testsFailed);
        await putMetric('TestDuration', totalDuration, StandardUnit.Milliseconds);
        
        // Summary
        log('=== Test Summary ===');
        log(`Total Tests: ${testsPassed + testsFailed}`);
        log(`Passed: ${testsPassed}`);
        log(`Failed: ${testsFailed}`);
        log(`Duration: ${totalDuration}ms`);
        
        if (testsFailed > 0) {
            const error = new Error(`${testsFailed} test(s) failed`);
            (error as any).responseBody = {
                success: false,
                testsPassed,
                testsFailed,
                duration: totalDuration,
                message: `${testsFailed} test(s) failed`
            };
            throw error;
        }
        
        return {
            success: true,
            testsPassed,
            testsFailed,
            duration: totalDuration,
            message: 'All tests passed'
        };
    } catch (error) {
        log(`Test suite error: ${error instanceof Error ? error.message : String(error)}`);
        
        // Ensure metrics are sent even on error
        const totalDuration = Date.now() - startTime;
        await putMetric('TestsFailed', testsFailed || 1);
        await putMetric('TestDuration', totalDuration, 'Milliseconds');
        
        throw error;
    }
}

File: lib/near/cdk.out/asset.cf0ef0dae1563dd3bac6960245525ef4d93606ec92d097f9162e0e7675c39b99/handler.py
#!/usr/bin/env python3
"""
Lambda handler for NEAR localnet test suite
Uses near-cli for all test operations
"""

import os
import json
import subprocess
import time
import boto3
from typing import Dict, Any

# CloudWatch client for metrics
cloudwatch = boto3.client('cloudwatch')
METRIC_NAMESPACE = "NEAR/Test"

# Test results tracking
tests_passed = 0
tests_failed = 0
test_start_time = int(time.time() * 1000)


def log(message: str):
    """Log message with timestamp"""
    print(f"[{time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}] {message}")


def put_metric(metric_name: str, value: float, unit: str = "Count"):
    """Send CloudWatch metric"""
    try:
        cloudwatch.put_metric_data(
            Namespace=METRIC_NAMESPACE,
            MetricData=[
                {
                    'MetricName': metric_name,
                    'Value': value,
                    'Unit': unit,
                    'Dimensions': [
                        {
                            'Name': 'FunctionName',
                            'Value': os.environ.get('AWS_LAMBDA_FUNCTION_NAME', 'near-localnet-test')
                        }
                    ]
                }
            ]
        )
    except Exception as e:
        log(f"Warning: Failed to put metric {metric_name}: {e}")


def run_test(test_name: str, command: list) -> bool:
    """Run a test command and return True if passed"""
    global tests_passed, tests_failed
    
    log(f"Running test: {test_name}")
    start_time = int(time.time() * 1000)
    
    try:
        result = subprocess.run(
            command,
            capture_output=True,
            text=True,
            timeout=120,  # 2 minute timeout per test
            check=True
        )
        end_time = int(time.time() * 1000)
        duration = end_time - start_time
        tests_passed += 1
        log(f"✅ PASS: {test_name} ({duration}ms)")
        if result.stdout:
            log(f"Output: {result.stdout[:500]}")  # First 500 chars
        return True
    except subprocess.TimeoutExpired:
        end_time = int(time.time() * 1000)
        duration = end_time - start_time
        tests_failed += 1
        log(f"❌ FAIL: {test_name} (timeout after {duration}ms)")
        return False
    except subprocess.CalledProcessError as e:
        end_time = int(time.time() * 1000)
        duration = end_time - start_time
        tests_failed += 1
        log(f"❌ FAIL: {test_name} ({duration}ms)")
        log(f"Error: {e.stderr[:500] if e.stderr else str(e)}")
        return False
    except Exception as e:
        end_time = int(time.time() * 1000)
        duration = end_time - start_time
        tests_failed += 1
        log(f"❌ FAIL: {test_name} ({duration}ms)")
        log(f"Exception: {str(e)}")
        return False


def check_near_cli() -> str:
    """Check if near-cli is available and return path"""
    # Check common locations
    possible_paths = [
        "/opt/near-cli/bin/near",
        "/var/task/near",
        "/usr/local/bin/near",
        "near"  # In PATH
    ]
    
    for path in possible_paths:
        try:
            result = subprocess.run(
                [path, "--version"],
                capture_output=True,
                timeout=5,
                check=True
            )
            log(f"Found near-cli: {path}")
            log(f"Version: {result.stdout.decode().strip()}")
            return path
        except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
            continue
    
    # If not found, try to use curl/jq as fallback for basic tests
    log("WARNING: near-cli not found, will use curl/jq fallback")
    return None


def run_read_tests(rpc_url: str, near_cli_path: str, test_depth: str):
    """Run read-only tests"""
    global tests_passed, tests_failed
    
    log("=== Running Read-Only Tests ===")
    
    if near_cli_path:
        # Test 1: RPC Status
        run_test("RPC Status", [near_cli_path, "--nodeUrl", rpc_url, "status"])
        
        # Test 2: Latest Block
        run_test("Latest Block", [near_cli_path, "--nodeUrl", rpc_url, "block", "latest"])
        
        # Test 3: View Account (system account)
        run_test("View System Account", [near_cli_path, "--nodeUrl", rpc_url, "view-account", "node0"])
        
        # Test 4: Chain Info (if available)
        if test_depth == "comprehensive":
            run_test("Chain Info", [near_cli_path, "--nodeUrl", rpc_url, "view-state"])
    else:
        # Fallback: Use curl for basic RPC tests
        import urllib.request
        
        # Test 1: RPC Status (curl)
        def curl_test(name: str, url: str):
            try:
                with urllib.request.urlopen(url, timeout=10) as response:
                    data = json.loads(response.read())
                    log(f"✅ PASS: {name}")
                    tests_passed += 1
                    return True
            except Exception as e:
                log(f"❌ FAIL: {name} - {str(e)}")
                tests_failed += 1
                return False
        
        curl_test("RPC Status", f"{rpc_url}/status")
        curl_test("Network Info", f"{rpc_url}/network_info")
    
    log(f"Read tests completed: {tests_passed} passed, {tests_failed} failed")


def run_write_tests(rpc_url: str, near_cli_path: str, test_depth: str, include_write: bool):
    """Run write tests (conditional)"""
    global tests_passed, tests_failed
    
    if not include_write:
        log("Skipping write tests (INCLUDE_WRITE_TESTS=false)")
        return
    
    if not near_cli_path:
        log("WARNING: Cannot run write tests without near-cli")
        return
    
    log("=== Running Write Tests ===")
    
    # Generate unique test account name
    test_account = f"test-{int(time.time())}.testnet"
    
    # Test 1: Create Account
    run_test("Create Test Account", [
        near_cli_path, "--nodeUrl", rpc_url,
        "create-account", test_account,
        "--masterAccount", "node0"
    ])
    
    # Test 2: Send Transaction (if account creation succeeded)
    if tests_failed == 0:
        run_test("Send Transaction", [
            near_cli_path, "--nodeUrl", rpc_url,
            "send", test_account, "node0", "1 NEAR"
        ])
    
    # Test 3: Deploy Contract (comprehensive mode only)
    if test_depth == "comprehensive" and tests_failed == 0:
        log("Skipping contract deployment (requires contract wasm file)")
        # Would need: run_test("Deploy Contract", [near_cli_path, "--nodeUrl", rpc_url, "deploy", test_account, "contract.wasm"])
    
    log(f"Write tests completed: {tests_passed} passed, {tests_failed} failed")


def lambda_handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """Lambda handler entry point"""
    global tests_passed, tests_failed, test_start_time
    
    # Reset counters
    tests_passed = 0
    tests_failed = 0
    test_start_time = int(time.time() * 1000)
    
    # Get configuration from environment or event
    rpc_url = os.environ.get("RPC_URL", "")
    network_id = os.environ.get("NETWORK_ID", "localnet")
    include_write_tests = os.environ.get("INCLUDE_WRITE_TESTS", "false").lower() == "true"
    test_depth = os.environ.get("TEST_DEPTH", "basic")
    instance_id = os.environ.get("INSTANCE_ID", "")
    
    # Override from event if provided
    if isinstance(event, dict):
        include_write_tests = event.get("includeWriteTests", include_write_tests)
        test_depth = event.get("testDepth", test_depth)
    
    log("=== NEAR Localnet Test Suite ===")
    log(f"RPC URL: {rpc_url}")
    log(f"Network ID: {network_id}")
    log(f"Include Write Tests: {include_write_tests}")
    log(f"Test Depth: {test_depth}")
    log(f"Instance ID: {instance_id}")
    
    # Validate RPC URL
    if not rpc_url:
        log("ERROR: RPC_URL environment variable not set")
        put_metric("TestsFailed", 1)
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "RPC_URL not set"})
        }
    
    # Measure RPC response time
    rpc_start = int(time.time() * 1000)
    try:
        import urllib.request
        with urllib.request.urlopen(f"{rpc_url}/status", timeout=5) as response:
            rpc_end = int(time.time() * 1000)
            rpc_duration = rpc_end - rpc_start
            put_metric("RpcResponseTime", rpc_duration, "Milliseconds")
            log(f"RPC response time: {rpc_duration}ms")
    except Exception as e:
        log(f"ERROR: Cannot reach RPC endpoint {rpc_url}: {e}")
        put_metric("TestsFailed", 1)
        return {
            "statusCode": 500,
            "body": json.dumps({"error": f"RPC endpoint unreachable: {str(e)}"})
        }
    
    # Check near-cli availability
    near_cli_path = check_near_cli()
    
    # Run read tests (always)
    run_read_tests(rpc_url, near_cli_path, test_depth)
    
    # Run write tests (conditional)
    run_write_tests(rpc_url, near_cli_path, test_depth, include_write_tests)
    
    # Calculate total duration
    test_end_time = int(time.time() * 1000)
    total_duration = test_end_time - test_start_time
    
    # Send metrics
    put_metric("TestsPassed", tests_passed)
    put_metric("TestsFailed", tests_failed)
    put_metric("TestDuration", total_duration, "Milliseconds")
    
    # Summary
    log("=== Test Summary ===")
    log(f"Total Tests: {tests_passed + tests_failed}")
    log(f"Passed: {tests_passed}")
    log(f"Failed: {tests_failed}")
    log(f"Duration: {total_duration}ms")
    
    # Prepare response
    response_body = {
        "success": tests_failed == 0,
        "testsPassed": tests_passed,
        "testsFailed": tests_failed,
        "duration": total_duration,
        "message": "All tests passed" if tests_failed == 0 else f"{tests_failed} test(s) failed"
    }
    
    # Return response
    if tests_failed > 0:
        log("❌ Test suite failed")
        # Raise exception to fail Lambda invocation (for Custom Resource)
        raise Exception(json.dumps(response_body))
    else:
        log("✅ All tests passed")
        return response_body

